lib/shared/s3-packages.js

```diff
@@ -0,0 +1,182 @@
+ * @type {String}
+ */
+const S3_PACKAGE_NAME = 'etcher';
+
+/**
+ * @summary Number of packages per Etcher version
+ * @constant
+ * @private
+ * @type {Number}
+ */
+const NUMBER_OF_PACKAGES = 8;
+
+/**
+ * @summary Find the latest version from an S3 bucket XML representation
+ * @function
+ * @private
+ *
+ * @param {String} bucketXML - s3 bucket XML
+ * @param {String} [range] - semver range
+ * @fulfil {String} - latest version
+ * @returns {Promise}
+ *
+ * request(s3Packages.BUCKET_URL_PRODUCTION, (error, response, body) => {
+ *   if (error) {
+ *     throw error;
+ *   }
+ *
+ *   s3Packages.getLatestVersionFromBucketXML(body, '>=2.5.0').then((latestVersion) => {
+ *     console.log(`The latest version is: ${latestVersion}`);
+ *   });
+ * });
+ */
+exports.getLatestVersionFromBucketXML = (bucketXML, range = '*') => {
+
+  /* eslint-disable lodash/prefer-lodash-method */
+
+  return xml.parseStringAsync(bucketXML)
+
+  /* eslint-enable lodash/prefer-lodash-method */
+
+    .get('ListBucketResult')
+    .get('Contents')
+    .reduce((accumulator, entry) => {
+      const [ name, version ] = _.split(_.first(entry.Key), '/');
+
+      // The `semver` module refuses to apply ranges to prerelease versions
+      // As a workaround, we drop the prerelease tags, if any, apply the range
+      // on that, and keep using the prerelease tag from then on.
+      // See https://github.com/npm/node-semver#prerelease-tags
+      const strippedVersion = `${semver.major(version)}.${semver.minor(version)}.${semver.patch(version)}`;
+
+      if (name === S3_PACKAGE_NAME && semver.satisfies(strippedVersion, range)) {
+        if (_.isNil(accumulator[version])) {
+          accumulator[version] = 1;
+        } else {
+          accumulator[version] += 1;
+        }
+
+        if (accumulator[version] >= NUMBER_OF_PACKAGES
+            && semver.gt(version, accumulator.latest)) {
+          accumulator.latest = version;
+        }
+      }
+
+      return accumulator;
+    }, {
+
+      // This is a good accumulator default value since
+      // every version is semantically greater than this.
+      latest: '0.0.0'
+
+    })
+    .get('latest');
+};
+
+/**
+ * @summary Query the latest available version from an S3 bucket
+ * @function
+ * @private
+ *
+ * @param {String} bucketUrl - an s3 bucket url
+ * @param {Object} options - options
+ * @param {String} options.currentVersion - current version
+ * @param {String} [options.range] - semver range
+ * @fulfil {String} - latest version
+ * @returns {Promise}
+ *
+ * @example
+ * s3Packages.queryLatestVersion(s3Packages.BUCKET_URL_PRODUCTION, {
+ *   currentVersion: '1.0.0-beta.16',
+ *   range: '<2.0.0'
+ * }).then((latestVersion) => {
+ *   console.log(`The latest version is: ${latestVersion}`);
+ * });
+ */
+exports.queryLatestVersion = (bucketUrl, options) => {
+  return request.getAsync(bucketUrl).then((response) => {
+    return exports.getLatestVersionFromBucketXML(response.body, options.range);
```

Comment:
I'm wondering if it might be worth making this module more granular, by having a memoized function which gets a list of all versions from a bucket URL, and then a separate function which takes that list of versions and gives you the 'latest' version, optionally filtering by a semver-range?

Something like the following pseudocode:
```
exports.getAllVersions = _.memoize((bucketUrl) => {
  // fetch URL, parse XML, return (possibly unordered) list of version-numbers
});

// this function could optionally be memoized too, but probably not worth it
exports.getLatestVersion = (bucketUrl, options) => {
  latestVersion = '0.0.0';
  for version in exports.getAllVersions(bucketUrl) {
    // get latest version, filtering by `options.range` if set
  }
  return latestVersion;
};
```
and that way you could (if you needed to) apply different semver-range filters at different times, without having to fetch the S3 URL again. (Although it has the obvious flaw that `getAllVersions` will only return versions for which there were `NUMBER_OF_PACKAGES` packages, but that's probably not something worth worrying about)